### PR TITLE
enable amp

### PR DESCRIPTION
**Pull request by hpasapp**

### CODE REVIEW

**Comment by hpasapp** [comment]
Note: this change requires torch 1.6.  I should probably do one of:
- make this take a parameter like `use_amp = False`, so that calling AMP autocast is not enabled by default, or
- check torch version before calling amp autocast

(I'm kind of leaning towards the latter, since this code doesnt really 'turn on' amp, so much as allow someone to turn it on, in their training loop)

**Comment by hpasapp** [comment]
- torchscript works again :)
- we only do amp things when in an autocast block
- added parameter amp_recurrence_fp16, to choose to use float16 for the recurrence kernel

Note that there is a very small change inside sru/cuda_functional.py  Not sure if there is a better way of doing this?

File: sru/ops.py [hunk 1]
```diff
@@ -89,25 +89,56 @@ def elementwise_recurrence_gpu(U: Tensor,
...
+            cast = torch.Tensor.half if amp_recurrence_fp16 else torch.Tensor.float
+
+            U = cast(U)
+            x = cast(x)
+            weight_c = cast(weight_c)
+            bias = cast(bias)
+            c_init = cast(c_init)
+            scale_x = cast(scale_x) if scale_x is not None else scale_x
+            dropout_mask_c = cast(dropout_mask_c) if dropout_mask_c is not None else dropout_mask_c
+            mask_pad = cast(mask_pad) if mask_pad is not None else mask_pad
```
Comment (taoleicn) [inline sru/ops.py:111]: dropout_mask_c should be cast as fp16 or fp32. but mask_pad should remain unchanged as byte / bool tensor

File: sru/cuda_functional.py [hunk 2]
```diff
@@ -64,6 +64,8 @@ def forward(ctx, u, x, weight_c, bias,
 
         forward_func = sru_cuda_lib.sru_bi_forward if bidirectional else \
             sru_cuda_lib.sru_forward
+        if h.dtype == torch.float16:
+            bias = bias.half()
```
Comment (taoleicn) [inline sru/cuda_functional.py:68]: One question: why do we need this conversion again given there are `cast()` calls in the `elementwise_recurrence_gpu()` method?

File: sru/ops.py (hunk 1 above)
Comment (taoleicn) [reply sru/ops.py:111]: Just curious: would `cast(tensor)` do nothing if `tensor` has already the dtype?

**Review by taoleicn — CHANGES_REQUESTED** [review]

File: sru/cuda_functional.py (hunk 2 above)
Comment (hpasapp) [reply sru/cuda_functional.py:68]: I dont know. Everything else arrives as float16. Except bias. For some reason. I guess I can try to dig into why a bit.

File: sru/ops.py (hunk 1 above)
Comment (hpasapp) [reply sru/ops.py:111]: Right. it becomes a `nop`. eg `float` is a `nop` if already `float32`:
Screen Shot 2020-10-03 at 1 44 52 PM

(from https://pytorch.org/docs/stable/amp.html )

File: sru/cuda_functional.py (hunk 2 above)
Comment (hpasapp) [reply sru/cuda_functional.py]: Hmmm, perhaps I only needed it when I entirely removed the autocast(disabled=True) block. And then I found eerything still worked iwth these two lines in. I think. anyway, I've removed these two lines now.

**Review by taoleicn — APPROVED** [review]

File: sru/cuda_functional.py (hunk 2 above)
Comment (taoleicn) [reply sru/cuda_functional.py]: Ok. sounds good to me now.

**Comment by taoleicn** [comment]
🎉

**Pull request merged**